In [1]:
import pandas as pd
from sklearn import linear_model
import statsmodels.api as sm

In [2]:
data = {'year': [2017,2017,2017,2017,2017,2017,2017,2017,2017,2017,2017,2017,2016,2016,2016,2016,2016,2016,2016,2016,2016,2016,2016,2016],
        'month': [12,11,10,9,8,7,6,5,4,3,2,1,12,11,10,9,8,7,6,5,4,3,2,1],
        'interest_rate': [2.75,2.5,2.5,2.5,2.5,2.5,2.5,2.25,2.25,2.25,2,2,2,1.75,1.75,1.75,1.75,1.75,1.75,1.75,1.75,1.75,1.75,1.75],
        'unemployment_rate': [5.3,5.3,5.3,5.3,5.4,5.6,5.5,5.5,5.5,5.6,5.7,5.9,6,5.9,5.8,6.1,6.2,6.1,6.1,6.1,5.9,6.2,6.2,6.1],
        'index_price': [1464,1394,1357,1293,1256,1254,1234,1195,1159,1167,1130,1075,1047,965,943,958,971,949,884,866,876,822,704,719]        
        }

In [3]:
df = pd.DataFrame(data)

In [4]:
x = df[['interest_rate','unemployment_rate']]
y = df['index_price']

In [5]:
regr = linear_model.LinearRegression()
regr.fit(x, y)

LinearRegression()

In [6]:
print('Intercept: \n', regr.intercept_)
print('Coefficients: \n', regr.coef_)

Intercept: 
 1798.403977625855
Coefficients: 
 [ 345.54008701 -250.14657137]


# Что мы нашли?
index_price = (intercept) + (interest_rate coef)*X1 + (unemployment_rate coef)*X2  
index_price = (1798.4040) + (345.5401)*X1 + (-250.1466)*X2

In [7]:
# Давай напечатааем всю стат модель
# with statsmodels
x = sm.add_constant(x) # adding a constant
 
model = sm.OLS(y, x).fit()
predictions = model.predict(x) 
 
print_model = model.summary()
print(print_model)

                            OLS Regression Results                            
Dep. Variable:            index_price   R-squared:                       0.898
Model:                            OLS   Adj. R-squared:                  0.888
Method:                 Least Squares   F-statistic:                     92.07
Date:                Sun, 07 May 2023   Prob (F-statistic):           4.04e-11
Time:                        17:18:47   Log-Likelihood:                -134.61
No. Observations:                  24   AIC:                             275.2
Df Residuals:                      21   BIC:                             278.8
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const              1798.4040    899.24

# Ploynomial regression

In [10]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
X = df[['interest_rate','unemployment_rate']]
y = df['index_price']

In [11]:
from sklearn.linear_model import LinearRegression
lin = LinearRegression()
  
lin.fit(X, y)

LinearRegression()

In [12]:
from sklearn.preprocessing import PolynomialFeatures
  
poly = PolynomialFeatures(degree = 4)
X_poly = poly.fit_transform(X)
  
poly.fit(X_poly, y)
lin2 = LinearRegression()
lin2.fit(X_poly, y)

LinearRegression()

In [13]:
print('Intercept: \n', lin2.intercept_)
print('Coefficients: \n', lin2.coef_)

Intercept: 
 1190978.148864942
Coefficients: 
 [ 0.00000000e+00  1.59078178e+06 -1.25908586e+06 -8.40860406e+05
 -2.33238929e+05  3.37091365e+05  1.60907859e+05  1.14507179e+05
  1.74069434e+03 -3.58652984e+04 -8.69804321e+03 -1.50355097e+04
 -1.67530239e+03  1.53688840e+02  1.41350503e+03]
